File: search-a-2d-matrix.cpp

// Time Complexity :O(n+m)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes

// Three line explanation of solution in plain english
// iterate over the array using two pointers(one at top right and the other at top left)
// if target is less than the element at current position,decrement the column pointer else incremnet the row pointer


class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if(matrix.empty()) {return false;}
        
        int row = matrix.size();
        int col= matrix[0].size();
        
        int low =0; 
        int high = row*col-1;
        while(low <= high){
            int mid = low + (high-low)/2;
            int e = matrix[mid/col][mid%col];
            if(e== target){
                return true;
            }if(e> target){
                high=mid-1;
            }else{
                low =mid+1;
            }
        }
        return false;
    }
};

File: search a 2d matrix.py

# // Time Complexity :O(m+n)
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode :Yes
# // Three line explanation of solution in plain english
#
# // Your code here along with comments explaining your approach

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])
        i = 0
        j = n-1
        while i<m and j>=0:
            if matrix[i][j]== target:
                return True
            if matrix[i][j]>target:
                j-=1
            else:
                i+=1
        return False